### DISCUSSION TITLE
Automatic Extension Install/Load

### DISCUSSION THREAD

**Discussion by pdet** [duckdb/duckdb#4607] [Ideas]
I think that installing and loading extensions is a slightly cumbersome process, and requires a bit of internal knowledge of the duckdb project (i.e., How to install/load, which extensions are available, ...)

I think we can add a configuration for automatic extension installation and loading.

Currently, all the extensions we support are either named functions or scanners, and either accessed by API wrappers or by SQL directly. 

We can add function definitions to the catalog and during execution check if the respective extension is already installed and loaded, if not do this process. The disadvantage is that whatever is being executed first will pay the cost of the download/installation but that sounds more desirable than throwing a ```function not found did you mean this...``` error.

Does anyone have a different idea/battle plan?

**Comment by hannes** [comment]
I do lets discuss later

**Comment by pdet** [comment]
Alright, so after a discussion with Hannes the battle plan will consist of:
1) Automatically detect extensions' functions.
On a CI job:
Build DuckDB without any extensions, and then for all extension types build them and load them into duckdb. After this, we can export which extensions add which functions to a config file.

2) Ship the configuration file respective to each DuckDB version

3) When looking up the catalog before executing a function, if the function is not in the catalog, also check the loaded config file.

4) If function exists in config file, download, install and load respective extension.

@Mytherin, maybe you would like to add something to this?

**Reply by Mytherin** [reply]
Automatic downloading and installing seems a bit much - perhaps just using this information to throw a better exception message (e.g. this function is provided by the X extension, but the extension is not loaded. Install and load the extension to use this function) is fine at least as a first pass here.

**Reply by Alex-Monahan** [reply]
In a corporate environment, these installs will also fail until we support using the http_proxy and https_proxy environment variables. #3836
